Manage Opkg configuration with toltecctl

matteodelabre opened the pull request:
This is a three-headed monster of a PR, addressing three distinct yet linked issues.

Features
=======

New toltecctl features
----------------------

At its core, this PR extends the toltecctl script (provided by toltec-bootstrap, introduced in #356) and enables it to manage the Opkg configuration, i.e., switch branches and migrate the configuration when needed. To achieve this without risking to overwrite custom config directives added by users, a split configuration approach is taken (as suggested by @Eeems):

* The main `/opt/etc/opkg.conf` file is automatically generated by concatenating the files in the `/opt/etc/opkg.conf.d/` folder (in lexicographic order).
* The file `/opt/etc/opkg.conf.d/10-entware.conf` configures the Entware repository and architecture names.
* The file `/opt/etc/opkg.conf.d/15-toltec.conf` configures the Toltec repository and architecture names.
* Users may create new files in `/opt/etc/opkg.conf.d` to add their custom directives.

New toltecctl commands are made available to users for managing the Opkg configuration:

* `toltecctl generate-opkg-conf` - Regenerates the main config file and automatically runs `opkg update` if the config changed
* `toltecctl switch-branch [BRANCH]` - Switches the current Toltec branch, rebuilds the config file and runs `opkg update`

This new `toltecctl switch-branch` fixes #341. A confirmation prompt is shown when someone tries to switch to the `testing` branch to warn them of the potential instability.

Bootstrap script revamp
-----------------------

This fixes #141.

With the previously mentioned additions, toltecctl now almost fully duplicates the features of the bootstrap script: it can create bind mounts, initialize the Opkg configuration, and define the user’s PATH. This PR therefore also contains a new bootstrap script that fetches the `toltec-bootstrap` package (containing toltecctl) first, and then uses its functions to perform the install, instead of reimplementing the same logic. By consolidating the features and variables of the two scripts, I hope to reduce the surface for potential bugs. This also makes it easier to add migrations to the `toltec-bootstrap` package, since it shares the code of the bootstrap script that’s used for new installs.

Migration to the new Opkg configuration
---------------------------------------

Building on the toltecctl improvements, a config migration logic was added to the `toltec-bootstrap` package’s configure function. This will automatically migrate single-opkg.conf-file installs to the new split config layout, while keeping the currently configured branch.

This migration is also able to transition configs using the old repository layout (with a single folder containing all packages) to the new arch-subfolder layout. By publishing the new toltec-bootstrap package to the root folder of the stable branch, this solves #370.

Test plan
==========

Should be tested both on rM1 and rM2.

Clean install
-------------

Starting with a clean system without a Toltec install, run the following command to test the new bootstrap script.

```console
$ wget https://toltec-dev.org/dev-manage-repos/bootstrap
$ echo "9195122984700c76ccdc58e25d09d0fca486324e8fc55ba781f6e1b812cc186c  bootstrap" | sha256sum -c
$ toltec_branch=dev-manage-repos bash bootstrap
```

<details>
<summary>Expected logs</summary>

```console
INFO:  Fetching secure wget
INFO:  Bootstrapping Opkg
2021-06-25 22:55:41 URL:https://bin.entware.net/armv7sf-k3.2/installer/opkg [752476/752476] -> "/home/root/.local/bin/opkg" [1]
2021-06-25 22:55:42 URL:https://toltec-dev.org/dev-manage-repos/rmall/toltec-bootstrap_0.1.0-1_rmall.ipk [5224/5224] -> "/tmp/toltec-bootstrap_0.1.0-1_rmall.ipk" [1]
Installing toltec-bootstrap (0.1.0-1) to root...
Configuring toltec-bootstrap.
INFO:  Installing Toltec and Entware
Mounting '/home/root/.entware' over '/opt'
Created symlink /etc/systemd/system/local-fs.target.wants/opt.mount → /lib/systemd/system/opt.mount.
Downloading https://bin.entware.net/armv7sf-k3.2/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/entware
Downloading https://toltec-dev.org/dev-manage-repos/rmall/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/toltec-rmall
Downloading https://toltec-dev.org/dev-manage-repos/rm2/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/toltec-rm2
Installing entware-opt (227000-3) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/entware-opt_227000-3_all.ipk
Installing libgcc (8.4.0-11) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/libgcc_8.4.0-11_armv7-3.2.ipk
Installing libc (2.27-11) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/libc_2.27-11_armv7-3.2.ipk
Installing libssp (8.4.0-11) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/libssp_8.4.0-11_armv7-3.2.ipk
Installing libpthread (2.27-11) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/libpthread_2.27-11_armv7-3.2.ipk
Installing librt (2.27-11) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/librt_2.27-11_armv7-3.2.ipk
Installing libstdcpp (8.4.0-11) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/libstdcpp_8.4.0-11_armv7-3.2.ipk
Installing entware-release (1.0-2) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/entware-release_1.0-2_all.ipk
Installing zoneinfo-asia (2021a-1) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/zoneinfo-asia_2021a-1_armv7-3.2.ipk
Installing zoneinfo-europe (2021a-1) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/zoneinfo-europe_2021a-1_armv7-3.2.ipk
Installing findutils (4.8.0-1) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/findutils_4.8.0-1_armv7-3.2.ipk
Installing terminfo (6.2-2) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/terminfo_6.2-2_armv7-3.2.ipk
Installing libpcre (8.44-4) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/libpcre_8.44-4_armv7-3.2.ipk
Installing grep (3.6-1a) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/grep_3.6-1a_armv7-3.2.ipk
Installing locales (2.27-9) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/locales_2.27-9_armv7-3.2.ipk
Installing opkg (2021-03-15-5936c4f9-1) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/opkg_2021-03-15-5936c4f9-1_armv7-3.2.ipk
Installing entware-upgrade (1.0-1) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/entware-upgrade_1.0-1_all.ipk
Configuring libgcc.
Configuring libc.
Configuring libssp.
Configuring libpthread.
Configuring librt.
Configuring terminfo.
Configuring libpcre.
Configuring grep.
Configuring locales.
Entware uses separate locale-archive file independent from main system
Creating locale archive /opt/usr/lib/locale/locale-archive
Adding en_EN.UTF-8
Adding ru_RU.UTF-8
You can download locale sources from http://bin.entware.net/other/i18n_glib227.tar.gz
You can add new locales to Entware using /opt/bin/localedef.new
Configuring entware-upgrade.
Upgrade operations are not required.
Configuring opkg.
Configuring zoneinfo-europe.
Configuring zoneinfo-asia.
Configuring libstdcpp.
Configuring entware-release.
Configuring findutils.
Configuring entware-opt.
Installing toltec-completion (0.2.0-1) to root...
Downloading https://toltec-dev.org/dev-manage-repos/rmall/toltec-completion_0.2.0-1_rmall.ipk
Installing bash-completion (2.11-1) to root...
Downloading https://toltec-dev.org/dev-manage-repos/rmall/bash-completion_2.11-1_rmall.ipk
Installing ca-certificates (20210119-1) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/ca-certificates_20210119-1_all.ipk
Installing wget-ssl (1.21.1-1) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/wget-ssl_1.21.1-1_armv7-3.2.ipk
Installing zlib (1.2.11-3) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/zlib_1.2.11-3_armv7-3.2.ipk
Installing libopenssl (1.1.1k-1) to root...
Downloading https://bin.entware.net/armv7sf-k3.2/libopenssl_1.1.1k-1_armv7-3.2.ipk
Configuring bash-completion.
Bash completions should take affect on next login.
To apply them immediately, run
 $ source /etc/profile.d/bash_completion.sh
Configuring ca-certificates.
Configuring zlib.
Configuring toltec-completion.
Configuring libopenssl.
Configuring wget-ssl.
Installing toltec-bootstrap (0.1.0-1) to root...
Downloading https://toltec-dev.org/dev-manage-repos/rmall/toltec-bootstrap_0.1.0-1_rmall.ipk
Configuring toltec-bootstrap.
Please restart your SSH session or run 'exec bash -l' to use Toltec
INFO:  After each system upgrade, run 'toltecctl reenable' to re-enable Toltec
```

</details>

After the install is complete, make sure `opkg` works (try to install some packages) and make sure `toltecctl` works (try running `toltecctl generate-opkg-conf`, which should be a no-op).

Migration from testing
----------------------

Starting with an existing Toltec install on the **testing** branch, edit your `/opt/etc/opkg.conf` to replace `testing` with `dev-manage-repos` (this will simulate the update reaching testing). Then, run the following commands:

```console
$ opkg update
$ opkg upgrade
```

<details>
<summary>Expected logs</summary>

```console
Upgrading toltec-bootstrap on root from 0.0.8-1 to 0.1.0-1...
Downloading https://toltec-dev.org/dev-manage-repos/rmall/toltec-bootstrap_0.1.0-1_rmall.ipk
Configuring toltec-bootstrap.
Please restart your SSH session or run 'exec bash -l' to use Toltec
Your Opkg configuration has been migrated
```

</details>

After upgrading, the Opkg configuration should have been migrated to use the new split format. Check that the branch name (`dev-manage-repos`) has been conserved and not replaced with the default `stable`. There should be two files called `10-entware.conf` and `15-toltec.conf` in `/opt/etc/opkg.conf.d`.

Migration from stable
---------------------

Starting with an existing Toltec install on the **stable** branch (ideally, with some packages already installed), edit your `/opt/etc/opkg.conf` to replace `stable` with `dev-manage-repos` (this will simulate the update reaching stable). Then, run the following commands:

```console
$ opkg update
$ opkg upgrade
$ opkg update
$ opkg upgrade
```

The first `upgrade` should trigger the migration to the new repository layout in the Opkg configuration. The second `upgrade` should upgrade your already-installed packages.

<details>
<summary>Expected logs</summary>

```console
$ opkg update
Downloading https://bin.entware.net/armv7sf-k3.2/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/entware
Downloading https://toltec-dev.org/dev-manage-repos/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/toltec
$ opkg upgrade
Upgrading toltec-bootstrap on root from 0.0.1-4 to 0.1.0-1...
Downloading https://toltec-dev.org/dev-manage-repos/toltec-bootstrap_0.1.0-1_armv7-3.2.ipk
Removing obsolete file /home/root/bootstrap.
Removing obsolete file /home/root/entware-reenable.
Configuring toltec-bootstrap.
Please restart your SSH session or run 'exec bash -l' to use Toltec
Your Opkg configuration has been migrated

! The upgrade is not finished! Please run `opkg update`
! and `opkg upgrade` once more to complete it.

$ opkg update
Downloading https://bin.entware.net/armv7sf-k3.2/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/entware
Downloading https://toltec-dev.org/dev-manage-repos/rmall/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/toltec-rmall
Downloading https://toltec-dev.org/dev-manage-repos/rm2/Packages.gz
Updated list of available packages in /opt/var/opkg-lists/toltec-rm2
$ opkg upgrade
Upgrading fret on root from 2.1.2~1 to 2.1.2-2...
Downloading https://toltec-dev.org/dev-manage-repos/rmall/fret_2.1.2-2_rmall.ipk
Installing display (1.0.1-5) to root...
Downloading https://toltec-dev.org/dev-manage-repos/rm2/display_1.0.1-5_rm2.ipk
Removing package rm2fb from root...
make sure to disable / uninstall any launchers like draft, oxide or remux before
rebooting your tablet to complete the uninstallation
Installing rm2fb-client (1.0.1-5) to root...
Downloading https://toltec-dev.org/dev-manage-repos/rm2/rm2fb-client_1.0.1-5_rm2.ipk
Upgrading tarnish on root from 2.1.2~1 to 2.1.2-2...
Downloading https://toltec-dev.org/dev-manage-repos/rmall/tarnish_2.1.2-2_rmall.ipk
Upgrading xochitl on root from 0.0.0-3 to 0.0.0-4...
Downloading https://toltec-dev.org/dev-manage-repos/rmall/xochitl_0.0.0-4_rmall.ipk
Multiple replacers for rm2fb, using first one (display).
display (1.0.1-5) already install on root.
Upgrading oxide on root from 2.1.2~1 to 2.1.2-2...
Downloading https://toltec-dev.org/dev-manage-repos/rmall/oxide_2.1.2-2_rmall.ipk
Upgrading rot on root from 2.1.2~1 to 2.1.2-2...
Downloading https://toltec-dev.org/dev-manage-repos/rmall/rot_2.1.2-2_rmall.ipk
Upgrading decay on root from 2.1.2~1 to 2.1.2-2...
Downloading https://toltec-dev.org/dev-manage-repos/rmall/decay_2.1.2-2_rmall.ipk
Upgrading erode on root from 2.1.2~1 to 2.1.2-2...
Downloading https://toltec-dev.org/dev-manage-repos/rmall/erode_2.1.2-2_rmall.ipk
rot (2.1.2-2) already install on root.
tarnish (2.1.2-2) already install on root.
decay (2.1.2-2) already install on root.
Upgrading remux on root from 0.1.8-1 to 0.1.9-2...
Downloading https://toltec-dev.org/dev-manage-repos/rmall/remux_0.1.9-2_rmall.ipk
erode (2.1.2-2) already install on root.
Configuring xochitl.
Configuring display.
Configuring rm2fb-client.
Configuring tarnish.
Configuring fret.
Configuring rot.
Configuring decay.
Configuring erode.
Configuring oxide.

Run the following command(s) to use oxide as your launcher
$ systemctl disable --now xochitl
$ systemctl enable --now tarnish

Configuring remux.

Run the following command(s) to use remux as your launcher
$ systemctl enable --now remux

```

</details>

Switching branches
------------------

After upgrading to this PR’s toltecctl (using one of the procedures above), try switching branches using `toltecctl switch-branch`. Trying to switch to the `testing` branch should display a prompt asking to confirm the switch. Try to add custom Opkg directives in a file in `/opt/etc/opkg.conf.d` and make sure they are conserved in the generated `/opt/etc/opkg.conf` even after switching branches.

matteodelabre committed:
Manage Opkg configuration with toltecctl

This is a three-headed monster of a PR, addressing three distinct yet linked issues.

New toltecctl features
----------------------

At its core, this PR extends the toltecctl script (provided by toltec-bootstrap, introduced in #356) and enables it to manage the Opkg configuration, i.e., switch branches and migrate the configuration when needed. To achieve this without risking to overwrite custom config directives added by users, a split configuration approach is taken (as suggested by @Eeems):

* The main `/opt/etc/opkg.conf` file is automatically generated by concatenating the files in the `/opt/etc/opkg.conf.d/` folder (in lexicographic order).
* The file `/opt/etc/opkg.conf.d/10-entware.conf` configures the Entware repository and architecture names.
* The file `/opt/etc/opkg.conf.d/15-toltec.conf` configures the Toltec repository and architecture names.
* Users may create new files in `/opt/etc/opkg.conf.d` to add their custom directives.

New toltecctl commands are made available to users for managing the Opkg configuration:

* `toltecctl generate-opkg-conf` - Regenerates the main config file and automatically runs `opkg update` if the config changed
* `toltecctl switch-branch [BRANCH]` - Switches the current Toltec branch, rebuilds the config file and runs `opkg update`

This new `toltecctl switch-branch` fixes #341. A confirmation prompt is shown when someone tries to switch to the `testing` branch to warn them of the potential instability.

Bootstrap script revamp
-----------------------

This fixes #141.

With the previously mentioned additions, toltecctl now almost fully duplicates the features of the bootstrap script: it can create bind mounts, initialize the Opkg configuration, and define the user’s PATH. This PR therefore also contains a new bootstrap script that fetches the `toltec-bootstrap` package (containing toltecctl) first, and then uses its functions to perform the install, instead of reimplementing the same logic. By consolidating the features and variables of the two scripts, I hope to reduce the surface for potential bugs. This also makes it easier to add migrations to the `toltec-bootstrap` package, since it shares the code of the bootstrap script that’s used for new installs.

Migration to the new Opkg configuration
---------------------------------------

Building on the toltecctl improvements, a config migration logic was added to the `toltec-bootstrap` package’s configure function. This will automatically migrate single-opkg.conf-file installs to the new split config layout, while keeping the currently configured branch.

This migration is also able to transition configs using the old repository layout (with a single folder containing all packages) to the new arch-subfolder layout. By publishing the new toltec-bootstrap package to the root folder of the stable branch, this will solve #370.

matteodelabre committed:
Fix migration regex to allow dashes in branch name

matteodelabre committed:
Fix Bash formatting

matteodelabre committed:
Merge branch 'testing' into package/toltec-bootstrap/manage-repos

matteodelabre commented:
(Only tested on rM2 for now, I will test on rM1 tomorrow.)

matteodelabre commented:
Tested on rM1 and rM2, version 2.8.0.

raisjn commented:
wow, awesome!

i tested on rM1 from testing -> dev-manage-repos (changed opkg.conf by hand)-> testing (using toltecctl script). the opkg upgrade messages were a bit hidden in my logs because i had a lot of packages updating.

matteodelabre commented:
Thank you for testing! Actually, it’s necessary to run `opkg update && opkg upgrade` twice only when you’re migrating from stable, not from testing. When migrating from stable, the first `opkg upgrade` only upgrades `toltec-bootstrap` and changes the configuration to point to the new subfolders and the second run upgrades the other packages. So the visibility of the message you mentioned should not be an issue for stable migrations. It’d be less confusing if that message wasn’t shown to users migrating from testing, so thanks for pointing it out!

matteodelabre committed:
Merge branch 'testing' into package/toltec-bootstrap/manage-repos

matteodelabre committed:
Only show `opkg upgrade` message for migrations from stable

matteodelabre committed:
Fix formatting

raisjn reviewed (dismissed):
shipit

seems good to me! maybe @LinusCDE or @Eeems have more review

LinusCDE commented on package/toltec-bootstrap/toltecctl:
```diff
@@ -169,48 +313,72 @@ uninstall-force() {
     reboot
```
```suggestion
    read -p "To complete the uninstall process, the device needs to be rebooted. Reboot now? [Y/n]" -n 1 -r
    echo

    if [[ $REPLY =~ ^[Nn]$ ]]; then
        echo "Please reboot your device manually to complete the uninstall process."
        return
     fi
    reboot
```

Tidbit: Not sure if user could get confused when the uninstall just reboots the device. Maybe add a prompt to ask the user to confirm it?

Eeems replied:
Perhaps also prompt to print the SSH password before reboot just in case they don't have it written down?

LinusCDE commented on package/toltec-bootstrap/toltecctl:
```diff
@@ -169,48 +313,72 @@ uninstall-force() {
+
+    case $action in
+        generate-opkg-conf)
+            generate-opkg-conf && opkg update
```
```suggestion
            generate-opkg-conf || return
            read -p "The config was updated. Do you want to apply it now ($ opkg update)? [Y/n]" -n 1 -r
            echo
            
            if [[ $REPLY =~ ^[Nn]$ ]]; then
                echo "Not updating for now."
                return
            fi
            
            opkg update
```

I'm not sure if `toltecctl generate-opkg-conf` should really do an update without asking (a user might want to verify the new config before it updates). Would be nice to add a prompt like suggested. Maybe even have some kind of `--no-update` flag.

matteodelabre replied:
Does this also apply to `toltecctl switch-branch`?

matteodelabre replied:
(I’m assuming it does.)

LinusCDE commented on package/toltec-bootstrap/toltecctl:
```diff
@@ -169,48 +313,72 @@ uninstall-force() {
+            uninstall
+            ;;
+
+        *)
```
```suggestion
        --help)
            help
            ;;
        *)
```
Make `toltecctl --help` not complain about an invalid command when it prints the help page anyways. Not sure whether we should include this in the usage as well.

LinusCDE commented on package/toltec-bootstrap/toltecctl:
```diff
@@ -169,48 +313,72 @@ uninstall-force() {
+
+        uninstall)
+            if [[ $# -eq 0 || ($1 != "-f" && $1 != "--force") ]]; then
+                read -p "This will wipe out all files in '$toltec_dest'. Continue? [y/n] " -n 1 -r
```
```suggestion
                read -p "This will wipe out all files in '$toltec_dest'. Continue? [y/N] " -n 1 -r
```

The implicit selection of "No" should be communicated with an upper case letter.

LinusCDE commented on package/toltec-bootstrap/toltecctl:
```diff
@@ -169,48 +313,72 @@ uninstall-force() {
+    <https://github.com/toltec-dev/toltec/blob/stable/docs/branches.md>
+
+MSG
+                read -p "Continue? [y/n] " -n 1 -r
```
```suggestion
                read -p "Continue? [y/N] " -n 1 -r
```

The implicit selection of "No" should be communicated with an upper case letter.

LinusCDE reviewed (dismissed):
I really like all of the concepts described!

Did most of your tests on both devices.

## RM2: Migrating from testing to dev-manage-repos

The migration worked perfectly! 

[Logs](https://haste.cosmos-ink.net/nijumiqyvo.txt)

## RM2: Clean install

Worked well! Did also a bit of testing with the new config directory. It updates as soon as new non-comment lines are added! It also seemed to strip out comments of my new `99-test.conf` file. Is this intentional?

[Logs](https://haste.cosmos-ink.net/ubatypiwut.txt)

I also tried adding a 2nd line comment to the file (and removed the duplicate `dest` configuration to ensure an update). It seems that comments from my `99-test.conf` don't get included at all, but the comment from the default ones do. **This may be some bug.**

## RM 1: Migrating stable to dev-manage-repos

Ran the steps after doing an uninstall and installing stable with remux, nao and retris.

The update worked fine. Maybe add `toltec-completion` (which adds `bash-completion`) during this migration since those are new in testing and would probably improve the user experience.

[Logs](https://haste.cosmos-ink.net/umejihijix.txt)

## RM1: Switching dev-manage-repos to testing with `switch-branch`

Worked well!

[Logs](https://haste.cosmos-ink.net/ezixesyral.txt)

## Some optional suggestions

I took the opportunity to also test some prompts. I added some optional suggestions. I don't see them needed, but they may improve the ux a bit.

matteodelabre commented:
Thanks all for the reviews and thanks @LinusCDE for the excellent suggestions!

Eeems committed:
Merge branch 'testing' into package/toltec-bootstrap/manage-repos

Eeems commented:
What's outstanding on this? It looks like we have two approvals, we can always open a new PR for anything else we want to change if there are no blockers.

matteodelabre commented:
It can be merged as-is, but Linus’ suggestions are minor enough that I think it makes more sense to handle them in this PR. I’m going to take care of them right now.

matteodelabre committed:
Indicate the default choice in yes/no prompts

Co-authored-by: Linus K. <linus@cosmos-ink.net>

matteodelabre committed:
Accept the help flag as a valid command

Co-authored-by: Linus K. <linus@cosmos-ink.net>

matteodelabre committed:
Add help to help message, also accept "help" and "-h"

matteodelabre commented:
> Worked well! Did also a bit of testing with the new config directory. It updates as soon as new non-comment lines are added! It also seemed to strip out comments of my new `99-test.conf` file. Is this intentional?
> 
> [Logs](https://haste.cosmos-ink.net/ubatypiwut.txt)
> 
> I also tried adding a 2nd line comment to the file (and removed the duplicate `dest` configuration to ensure an update). It seems that comments from my `99-test.conf` don't get included at all, but the comment from the default ones do. **This may be some bug.**

@LinusCDE This is intentional, but since it’s confused you I’m not sure it’s the best behavior. The intent was to avoid drowning the comments that indicate which file each directive block comes from with comments from the original files.

matteodelabre committed:
Prompt before rebooting after uninstall + show SSH password

matteodelabre commented:
> Maybe add `toltec-completion` (which adds `bash-completion`) during this migration since those are new in testing and would probably improve the user experience.

Unfortunately we can’t call `opkg` while inside of a maintainer script like the migration script. This is also the reason why, when migrating from stable, users need to call `opkg update && opkg upgrade` twice manually. We could add `toltec-completion` as a dependency of `toltec-bootstrap`, but that would prevent users from uninstalling `toltec-completion` if they don’t want/need them.

matteodelabre committed:
Merge branch 'testing' into package/toltec-bootstrap/manage-repos

matteodelabre committed:
Prompt user before running `opkg update` in `switch-branch` or `generate-opkg-conf`

matteodelabre commented:
I think I’ve addressed all your comments @LinusCDE. I’ve updated the dev repo at <https://toltec-dev.org/dev-manage-repos> with the changes. I’ll do more thorough testing tomorrow.

matteodelabre committed:
Merge branch 'testing' into package/toltec-bootstrap/manage-repos

Eeems committed:
Merge branch 'testing' into package/toltec-bootstrap/manage-repos

matteodelabre committed:
switch-branch: Warn if the branch doesn’t exist

As suggested by @Eeems

matteodelabre committed:
Fix errexit triggered by one-liner flag check

matteodelabre committed:
Fix shellcheck error

matteodelabre committed:
bootstrap: Do not fail error function on cleanup issue

matteodelabre committed:
Update bootstrap checksum

matteodelabre committed:
bootstrap: Display the help message before attempting cleanup

This is to make sure that if a cleanup operation fails, the help message
is shown to the user anyway.

matteodelabre committed:
Add “help” to toltecctl completions

matteodelabre commented:
Performed additional testing on rM1 and rM2. I think this is clear for merging in testing.

Eeems reviewed (approved):
lgtm
